File: LeetCode/199-binary-tree-right-side-view/Solution.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Computes the right-side view of a binary tree using BFS. Each level is traversed left-to-right,
 * and the last node encountered at that level is recorded as the visible node.
 */
class Solution {

  /**
   * Returns the list of node values visible from the right side.
   *
   * @param root the root of the binary tree
   * @return a list of integers representing the rightmost node at each depth
   */
  public List<Integer> rightSideView(TreeNode root) {
    // Result list storing the rightmost node per level
    List<Integer> rightside = new ArrayList<>();
    if (root == null)
      return rightside;

    // Queue holding nodes for the next level of BFS
    Queue<TreeNode> nextLevel = new LinkedList<>();
    nextLevel.add(root);

    // Process the tree level by level
    while (!nextLevel.isEmpty()) {
      // Move all nodes from nextLevel into currLevel
      Queue<TreeNode> currLevel = nextLevel;
      nextLevel = new LinkedList<>();

      TreeNode node = null; // Tracks the last node processed at this level

      // Traverse all nodes in the current level
      while (!currLevel.isEmpty()) {
        node = currLevel.poll();

        // Enqueue children for the next level
        if (node.left != null) {
          nextLevel.add(node.left);
        }
        if (node.right != null) {
          nextLevel.add(node.right);
        }
      }

      // node now holds the rightmost node of this level
      if (node != null) {
        rightside.add(node.val);
      }
    }

    return rightside;
  }
}

File: LeetCode/199-binary-tree-right-side-view/solution.ts
/**
 * Represents a node in a binary tree.
 */
class TreeNode {
  /** The integer value stored in the node. */
  val: number;

  /** Reference to the left child node, or null if absent. */
  left: TreeNode | null;

  /** Reference to the right child node, or null if absent. */
  right: TreeNode | null;

  /**
   * Creates a new TreeNode instance.
   *
   * @param val - The node's value. Defaults to 0.
   * @param left - The left child node. Defaults to null.
   * @param right - The right child node. Defaults to null.
   */
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * Computes the right-side view of a binary tree using BFS.
 * The right-side view consists of the last visible node at each depth level.
 *
 * @param root - The root of the binary tree.
 * @returns An array of node values representing the rightmost node at each level.
 */
function rightSideView(root: TreeNode | null): number[] {
  const rightside: number[] = [];
  if (root === null) return rightside;

  // Queue of nodes to process for the next depth level
  let nextLevel: TreeNode[] = [root];

  while (nextLevel.length > 0) {
    const currLevel = nextLevel;
    nextLevel = [];
    let node: TreeNode | null = null;

    // Traverse all nodes in the current level
    for (let i = 0; i < currLevel.length; i++) {
      node = currLevel[i];

      if (node.left) {
        nextLevel.push(node.left);
      }
      if (node.right) {
        nextLevel.push(node.right);
      }
    }

    // node now holds the rightmost node of this level
    if (node) {
      rightside.push(node.val);
    }
  }

  return rightside;
}